Mathlib/Algebra/SkewMonoidAlgebra/Basic.lean

```diff
@@ -0,0 +1,1243 @@
+/-
+Copyright (c) 2024 María Inés de Frutos Fernández. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: María Inés de Frutos Fernández, Xavier Généreux
+-/
+import Mathlib.Algebra.Algebra.Equiv
+import Mathlib.Algebra.Algebra.NonUnitalHom
+import Mathlib.Data.Finsupp.Basic
+import Mathlib.LinearAlgebra.FreeModule.Basic
+
+/-!
+
+This file presents a skewed version of `Mathlib.Algebra.MonoidAlgebra.Basic`.
+
+We define `SkewMonoidAlgebra k G` attached with a skewed convolution product.
+Here, the product of two elements `f g : SkewMonoidAlgebra k G` is the finitely supported
+function whose value at `a` is the sum of `f x * (x • g y)` over all pairs `x, y`
+such that `x * y = a`.
+
+The associativity of the skewed multiplication depends on the `[MulSemiringAction G k]` instance.
+In particular, this means that unlike in `Mathlib.Algebra.MonoidAlgebra.Basic`, `G` will
+need to be a monoid for most of our uses.
+-/
+
+noncomputable section
+
+/-- The skew monoid algebra over a semiring `k` generated by the monoid `G`.
+It is the type of finite formal `k`-linear combinations of terms of `G`,
+endowed with a skewed convolution product. -/
+structure SkewMonoidAlgebra (k : Type*) (G : Type*) [Zero k] where
+  /-- Map **from** `G →₀ k`. -/
+  ofFinsupp ::
+  /-- Map **to** `G →₀ k`. -/
+  toFinsupp : G →₀ k
+
+open Function hiding Commute
+namespace SkewMonoidAlgebra
+
+universe u₁ u₂
+
+variable {k : Type u₁} {G : Type u₂} {H : Type*} {R : Type*}
+
+section AddCommMonoid
+
+variable [AddCommMonoid k]
+
+@[simp]
+theorem eta (f : SkewMonoidAlgebra k G) : ofFinsupp f.toFinsupp = f := by cases f; rfl
+
+private irreducible_def add :
+    SkewMonoidAlgebra k G → SkewMonoidAlgebra k G → SkewMonoidAlgebra k G
+  | ⟨a⟩, ⟨b⟩ => ⟨a + b⟩
+
+private irreducible_def smul {S : Type*} [SMulZeroClass S k] :
+    S → SkewMonoidAlgebra k G → SkewMonoidAlgebra k G
+  | s, ⟨b⟩ => ⟨s • b⟩
```

Comment:
`smul` can't be irreducible, as this creates diamonds. I think there is a commend about this for `Polynomial`.

Reply:
@YaelDillies Is it OK with you if I revert the change?

Reply:
Do we want to hide these from the kernel and frontend or just the frontend? I generally prefer `irreducible` over `irreducible_def` unless we really see an effect on the kernel. 

I worry that the diamonds are just a indication of poor API boundaries for current implementation of `Polynomial`

Reply:
The problem is that making `smul` irreducible makes it impossible for `smul a p` and `smul (f a) p` to be defeq, even if the former contains `f` internally.